modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/pii/PiiPseudonymizerEnrichment.scala

```diff
@@ -132,6 +133,37 @@ object PiiPseudonymizerEnrichment extends ParseableEnrichment {
       .get(fieldName)
       .map(_.asRight)
       .getOrElse(s"The specified json field $fieldName is not supported".asLeft)
+
+  /** Helper to remove fields that were wrongly added and are not in the original JSON. See #351. */
+  private[pii] def removeAddedFields(hashed: Json, original: Json): Json =
+    hashed.asObject.map(_.toMap) match {
+      case Some(hashed_fields) => // hashed is JSON object
+        original.asObject.map(_.toMap) match {
+          case Some(orig_fields) =>
+            val newMap =
+              hashed_fields
+                .collect {
+                  case (k, v) if orig_fields.isDefinedAt(k) =>
+                    (k, removeAddedFields(v, orig_fields.get(k).get))
+                }
+            Json.fromFields(newMap)
+          case None =>
+            hashed // should never happen. Would mean change of type of one field
+        }
+      case None =>
+        hashed.asArray match {
+          case Some(hashed_arr) => // hashed is array (can contain JSON objects)
+            original.asArray match {
+              case Some(orig_arr) =>
+                // we can use zip because there should never be new fields in an array, only in an object
+                val values = hashed_arr.zip(orig_arr).map { case (hashed, orig) => removeAddedFields(hashed, orig) }
+                Json.fromValues(values)
+              case None =>
+                Json.fromValues(hashed_arr) // should never happen. Would mean change of type of one field
+            }
+          case None => hashed // hashed is neither JSON object nor array
+        }
+    }
```

Comment:
My modest implementation:

```suggestion
  private[pii] def removeAddedFields(hashed: Json, original: Json): Json = {
    val fixedObject = for {
      hashedFields <- hashed.as[Map[String, Json]]
      originalFields <- original.as[Map[String, Json]]
      newMap = hashedFields.flatMap {
        case (k, v) => originalFields.get(k) match {
          case Some(originalValue) => Some((k, removeAddedFields(v, originalValue)))
          case None => None
        }
      }
    } yield Json.fromFields(newMap)

    lazy val fixedArray = for {
      hashedArr <- hashed.asArray
      originalArr <- original.asArray
      newArr = hashedArr.zip(originalArr).map { case (hashed, orig) => removeAddedFields(hashed, orig) }
    } yield Json.fromValues(newArr)

    fixedObject.toOption.orElse(fixedArray).getOrElse(hashed)
  }
```

Reply:
I'd personally replace `flatMap` on 142 with `unite` as I hate `A => Option[B]` flatMapping on `List[A]`, but it will require additional cats import.

Reply:
Thanks for the better implementation @chuwy !